profiling/src/string_table/owned.rs

```diff
@@ -0,0 +1,109 @@
+use super::bump_owned::BorrowedStringTable;
+use bumpalo::Bump;
+use self_cell::self_cell;
+use std::ops::Range;
+
+// todo: use ouroboros instead?
+self_cell!(
+    struct StringTableCell {
+        owner: Bump,
+
+        #[covariant]
+        dependent: BorrowedStringTable,
+    }
+);
+
+pub struct OwnedStringTable {
+    inner: StringTableCell,
+}
+
+impl Clone for OwnedStringTable {
+    fn clone(&self) -> Self {
+        let bytes = self
+            .inner
+            .with_dependent(|arena, _table| arena.allocated_bytes());
+
+        use super::StringTable as StringTableTrait;
+        let mut table = OwnedStringTable::with_capacity(bytes);
+        let len = self.len();
+        table.reserve(len);
+        for str in self.get_range(0..len) {
+            table.insert(str);
+        }
+        table
+    }
+}
+
+impl OwnedStringTable {
+    #[inline]
+    pub fn new() -> Self {
+        Self::with_capacity(4000)
+    }
+
+    #[inline]
+    pub fn with_capacity(capacity: usize) -> Self {
+        let inner = StringTableCell::new(Bump::with_capacity(capacity), |arena| {
+            BorrowedStringTable::new(arena)
+        });
+        Self { inner }
+    }
+
+    #[inline]
+    fn reserve(&mut self, additional: usize) {
+        self.inner.with_dependent_mut(|_arena, table| {
+            table.set.vec.reserve(additional);
+            table.set.map.reserve(additional);
+        })
+    }
+}
+
+impl Default for OwnedStringTable {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl super::StringTable for OwnedStringTable {
+    #[inline]
+    fn len(&self) -> usize {
+        self.inner.with_dependent(|_arena, set| set.len())
+    }
+
+    #[inline]
+    fn insert_full(&mut self, str: &str) -> (usize, bool) {
+        self.inner
+            .with_dependent_mut(|_arena, set| set.insert_full(str))
```

Comment:
@morrisonlevi could you help me understand what `with_dependent_mut` and `with_dependent` are doing? I could not find any kind of documentation on this

Reply:
This is part of the `self_cell` crate:
https://docs.rs/self_cell/latest/self_cell/macro.self_cell.html
Basically, they are borrowing the "dependent" which is the borrowed string table for the duration of the closure. This is how they scope the lifetime of the borrow for safety.
These things are needed because self-referential structs (structs which internally have references to other parts of the same struct), are currently not supported in safe Rust.